src/server.cpp

```diff
@@ -1603,49 +1606,77 @@ void Server::SendShowFormspecMessage(session_t peer_id, const std::string &forms
 	Send(&pkt);
 }
 
-// Spawns a particle on peer with peer_id
-void Server::SendSpawnParticle(session_t peer_id, u16 protocol_version,
-	const ParticleParameters &p)
+void Server::SendSpawnParticles(session_t peer_id, const std::vector<ParticleParameters> &particles)
 {
 	static thread_local const float radius =
-			g_settings->getS16("max_block_send_distance") * MAP_BLOCKSIZE * BS;
-
-	if (peer_id == PEER_ID_INEXISTENT) {
-		std::vector<session_t> clients = m_clients.getClientIDs();
-		const v3f pos = p.pos * BS;
-		const float radius_sq = radius * radius;
-
-		for (const session_t client_id : clients) {
-			RemotePlayer *player = m_env->getPlayer(client_id);
-			if (!player)
-				continue;
+		g_settings->getS16("max_block_send_distance") * MAP_BLOCKSIZE * BS;
+	const float radius_sq = radius * radius;
 
-			PlayerSAO *sao = player->getPlayerSAO();
-			if (!sao)
-				continue;
+	RemoteClient *client = m_clients.getClientNoEx(peer_id, CS_Active);
+	if (!client)
+		return;
 
-			// Do not send to distant clients
-			if (sao->getBasePosition().getDistanceFromSQ(pos) > radius_sq)
-				continue;
+	PlayerSAO *sao = getPlayerSAO(peer_id);
+	if (!sao)
+		return;
 
-			SendSpawnParticle(client_id, player->protocol_version, p);
+	RemotePlayer *player = m_env->getPlayer(peer_id);
```

Comment:
Might need a `EnvAutoLock envlock(this);` because `ServerEnvironment::getPlayer` is not thread-safe.

Reply:
I'll take a look

Reply:
Looks like I should do that. I understand that `ServerEnvironment::(add|remove)Player` may modify the env's `m_players`, coming from the network thread, which acquires the envlock in `Server::ProcessData`. 

Done